cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/AsyncLoopSteps.java

```diff
@@ -0,0 +1,219 @@
+                branchContext.putVariable(asyncLoopStatement.getVarName(), splitItem);
+
+                updateCallArgumentsAndPushContextToStack(branchRuntimeEnvironment, branchContext, new HashMap<String, Serializable>());
+
+                createBranch(
+                        branchRuntimeEnvironment,
+                        executionRuntimeServices,
+                        RUNNING_EXECUTION_PLAN_ID,
+                        refId,
+                        nextStepId,
+                        branchBeginStep);
+
+                fireEvent(
+                        executionRuntimeServices,
+                        runEnv,
+                        ScoreLangConstants.EVENT_BRANCH_START,
+                        "async loop branch created",
+                        Pair.of(ScoreLangConstants.REF_ID, refId),
+                        Pair.of(ScoreLangConstants.SPLIT_ITEM_KEY, splitItem),
+                        Pair.of(LanguageEventData.levelName.TASK_NAME.name(), nodeName));
+            }
+
+            updateCallArgumentsAndPushContextToStack(runEnv, flowContext, new HashMap<String, Serializable>());
+
+            // forward after the branches are created because begin task method also calls forward
+            runEnv.getExecutionPath().forward();
+        } catch (RuntimeException e) {
+            logger.error("There was an error running the add branches execution step of: \'" + nodeName + "\'. Error is: " + e.getMessage());
+            throw new RuntimeException("Error running: " + nodeName + ": " + e.getMessage(), e);
+        }
+
+    }
+
+    public void joinBranches(@Param(ScoreLangConstants.RUN_ENV) RunEnvironment runEnv,
+                             @Param(EXECUTION_RUNTIME_SERVICES) ExecutionRuntimeServices executionRuntimeServices,
+                             @Param(ScoreLangConstants.TASK_AGGREGATE_KEY) List<Output> taskAggregateValues,
+                             @Param(ScoreLangConstants.TASK_NAVIGATION_KEY) Map<String, ResultNavigation> taskNavigationValues,
+                             @Param(ScoreLangConstants.NODE_NAME_KEY) String nodeName) {
+        try {
+            List<EndBranchDataContainer> branches = executionRuntimeServices.getFinishedChildBranchesData();
+            List<Map<String, Serializable>> branches_context = Lists.newArrayList();
+            Context flowContext = runEnv.getStack().popContext();
+            Map<String, Serializable> contextBeforeSplit = flowContext.getImmutableViewOfVariables();
+            List<String> branchResults = Lists.newArrayList();
+
+            for (EndBranchDataContainer branch : branches) {
+                Map<String, Serializable> branchContext = branch.getContexts();
+
+                RunEnvironment branchRuntimeEnvironment = (RunEnvironment) branchContext.get(ScoreLangConstants.RUN_ENV);
+
+                branches_context.add(branchRuntimeEnvironment.getStack().popContext().getImmutableViewOfVariables());
```

Comment:
either branch_contexts and branchResults or brnaches_context and branchesResult...